pkg/aws/gather.go

```diff
@@ -0,0 +1,89 @@
+package aws
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/red-hat-storage/ocs-osd-deployer/utils"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html
+	IMDSv1Server      = "http://169.254.169.254"
+	DataConfigMapName = "aws-data"
+	CidrKey           = "vpc-cidr"
+)
+
+func GatherData(imdsServerAddr string, k8sClient client.Client, namespace string, log logr.Logger) error {
+	// A mac address for one of the interfaces on the instance is needed
+	// to query for the VPC CIDR block.
+	log.Info("Determining mac address on instance")
+	var mac string
+	err := utils.Retry(3, time.Second, func() error {
+		endpoint := fmt.Sprintf("%s/latest/meta-data/mac", imdsServerAddr)
+		resp, err := http.Get(endpoint)
+		if err != nil {
+			return fmt.Errorf("Failed to GET %s: %v", endpoint, err)
+		}
+		defer resp.Body.Close()
+
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return fmt.Errorf("Failed to read response body: %v", err)
+		}
+		mac = string(body)
+		return nil
+	})
```

Comment:
this code is repeating between the 2 get operations, this will be a good candidate for a refactor. You can introduce another func 
```
func getAndParseBody(endpoint) (string, error) {
	resp, err := http.Get(endpoint)
	if err != nil {
		return "", fmt.Errorf("Failed to GET %s: %v", endpoint, err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", fmt.Errorf("Failed to read response body: %v", err)
	}
	return string(body), nil
}
```
then 
```suggestion
	err := utils.Retry(3, time.Second, func() error {
		endpoint := fmt.Sprintf("%s/latest/meta-data/mac", imdsServerAddr)
		if mac, err = getAndParseBody(); err != nil {
			return err
		} 
	})
```

Reply:
Thanks a lot for this @nb-ohad , your suggested refactor significantly improves code readability. While looking at it, I had an idea for a code-review strategy that I wanted to run by you to understand the disadvantages of the approach. The following comment will be what I would add if I were the reviewer, let me know what you think:

Reply:
```suggestion
// TODO: Needs refactor
	err := utils.Retry(3, time.Second, func() error {
		endpoint := fmt.Sprintf("%s/latest/meta-data/mac", imdsServerAddr)
		resp, err := http.Get(endpoint)
		if err != nil {
			return fmt.Errorf("Failed to GET %s: %v", endpoint, err)
		}
		defer resp.Body.Close()

		body, err := io.ReadAll(resp.Body)
		if err != nil {
			return fmt.Errorf("Failed to read response body: %v", err)
		}
		mac = string(body)
		return nil
	})
```

Reply:
Here's my reasoning for this:

- It saves the reviewer from having to refactor the code for the contributor.
- It gives the contributor the opportunity to refactor the code themselves.
- It doesn't block merging the current PR.
- If the original contributor doesn't want to do the refactor, it gives other contributors a chance to do so.
- If no other contributor knows how to do the refactor, they can reach out to the reviewer that placed the suggestion there to learn how to do it.

I see it like a "teach a man to fish vs. giving the man a fish" scenario.

Reply:
@renan-campos 
As a reviewer, you identify places that might benefit from refactoring only when you have some notion of the refactoring that you would like to see. Putting a comment saying `Needs refactor` does not help the contributor to understand what kind of refactoring you have in mind, or even evaluate or rebuttal the refactoring request. 
In addition, my assumption is that the contributor didn't find the desired way to refactor the code because if he did, he would have submitted the refactored code in the first place. Just pointing to the fact that some piece of code need refactoring will not help the contributor to come up with a refactoring strategy. 

This all boils down to the conclusion that as a reviewer if you think a specific refactoring is needed - just propose it. And this is exactly the kind of review comments I tend to leave in these scenarios.

Reply:
Now as a maintainer I will say that you have the responsibility to gate code that does not meet a specific standard, so:
> It doesn't block merging the current PR.
> f the original contributor doesn't want to do the refactor, it gives other contributors a chance to do so.
These are not valid points, as this code will not be approved for merging without a proper rebuttal

Reply:
Isn't the point 
> * It doesn't block merging the current PR.
Valid since this suggestion blocks function for form? I might be wrong, but I think the code that is there and the code you suggested does the exact same thing, only yours does so more clearly?

Reply:
No, you are correct both forms are accomplishing the exact same thing and even the original code (before all refactors) did the exact same thing. There are infinite ways to achieve the exact same behavior.
But a review is not only about the correctness of a contribution, it is also about other properties of the contributions. Contributed code is measured by multiple factors:

* Correctness
* Performance - when relevant and across multiple dimensions 
* future maintenance cost
* Standards - which relates to the prev bullet
* Readability - which then again is measured across multiple dimensions 
* Reusability - when relevant
and so on. 

And of course some of these are subjective measurments - but this is why a maintainer on one project might not fit the position of a maintainer on a different project

Reply:
I see the value of the suggestion : improved readability reduces future maintenance cost.

I don't want to keep you from your weekend, I had one last thought on this if you want to talk more monday:
The factors that need to be taken into account when reviewing contributed code affect how the number of contributors and reviewers can scale. If the standards are too stringent or subjective:
- fewer contributors will be able to confidently and independently make changes to the code
- fewer other reviewers will be able to match the quality of reviews of the OG reviewer

In an extreme case, the long term result could be reviewers and contributors that can only code and review in exactly the same way as the original.

Reply:
I am not sure that I agree here. If I was to implement this feature I would have done it differently. All of the review suggestions are amounting to a couple of requested changes to some pieces of code that I find a little bit lacking. The Idea was never to push you to implement my vision for a solution. 

Saying that I do agree this entire discussion boils down to an exercise of balance. 
As you have suggested, let's continue this meta-discussion outside of the PR comments.